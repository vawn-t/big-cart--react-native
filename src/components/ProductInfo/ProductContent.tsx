import { View } from 'react-native';

// Types
import { FontWeight, SizeType, TypoVariant } from '@interfaces/index';

// Components
import { Typography } from '@components/common';

// Stores
import useStore from '@stores/useStore';

// Styles
import styles from './styles';

interface IProps {
  price: number;
  name: string;
  productUnitId: number;
  description: string;
}

const ProductContent = ({
  price,
  name,
  productUnitId,
  description,
}: IProps) => {
  // Stores
  const getProductUnitNameById = useStore.use.getProductUnitNameById();

  return (
    <>
      <View style={styles.header}>
        <Typography
          text={`$${price}`}
          style={styles.price}
          fontWeight={FontWeight.SemiBold}
        />
        <Typography
          text={name}
          size={SizeType.Large}
          fontWeight={FontWeight.SemiBold}
          variant={TypoVariant.Paragraph2}
        />
        <Typography
          text={getProductUnitNameById(productUnitId)}
          size={SizeType.Small}
          fontWeight={FontWeight.Medium}
        />
      </View>

      <Typography
        text={description}
        size={SizeType.Small}
        style={styles.description}
        numberOfLines={6}
      />
    </>
  );
};

export default ProductContent;
