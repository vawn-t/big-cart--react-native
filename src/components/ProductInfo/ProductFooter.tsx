import { memo, useCallback, useEffect, useState } from 'react';
import { View, TouchableOpacity } from 'react-native';
import Toast from 'react-native-root-toast';

// Components
import { Button, Typography } from '@components/common';
import { AddIcon, MinusIcon } from '@components/Icon';

// Types
import { FontWeight, IconName, SizeType, TypoVariant } from '@interfaces/index';

// Stores
import useStore from '@stores/useStore';

// Hooks
import useCartItemAdd from '@hooks/useCart/useCartItemAdd';
import useCartItemUpdate from '@hooks/useCart/useCartItemUpdate';

// Models
import { CartItem } from '@models/index';

// Constants
import { SUCCESS } from '@constants/messages';

// Styles
import styles from './styles';

interface IProps {
  productId: number;
}

const ProductFooter = ({ productId: productId }: IProps) => {
  const [cartItem, setCartItem] = useState<CartItem | null>(null);
  const [quantity, setQuantity] = useState(1);

  // Stores
  const cartItemById = useStore.use.cartItemById();
  const getCartItemByProductId = useStore.use.getCartItemByProductId();

  // Queries
  const {
    mutateAsync: addToCart,
    isPending: isAddCartPending,
    isSuccess: isAddCartItemSuccess,
    isError: isAddCartItemError,
    error: cartItemError,
  } = useCartItemAdd();

  const {
    mutateAsync: updateCartItem,
    isPending: isUpdateCartPending,
    isSuccess: isCartItemUpdated,
    error: updateCartItemError,
    isError: isUpdateCartItemError,
  } = useCartItemUpdate();

  const cartItemId = getCartItemByProductId(productId)?.id;

  useEffect(() => {
    if (isAddCartItemSuccess) {
      Toast.show(SUCCESS.ADD_TO_CART);
    } else if (isAddCartItemError) {
      if (cartItemError.response?.data.errors?.length) {
        Toast.show(cartItemError.response.data.errors[0].msg);
      } else {
        Toast.show(cartItemError.message);
      }
    }
  }, [isAddCartItemSuccess, isAddCartItemError, cartItemError]);

  useEffect(() => {
    if (cartItemId) {
      const cartItem = cartItemById[cartItemId];

      setCartItem(cartItem);
      setQuantity(cartItem.quantity);
    }

    return () => {
      setQuantity(1);
      setCartItem(null);
    };
  }, [cartItemId]);

  useEffect(() => {
    if (isCartItemUpdated) {
      Toast.show(SUCCESS.UPDATE_CART_ITEM);
    } else if (isUpdateCartItemError) {
      if (updateCartItemError.response?.data.errors?.length) {
        Toast.show(updateCartItemError.response.data.errors[0].msg);
      } else {
        Toast.show(updateCartItemError.message);
      }
    }
  }, [isCartItemUpdated, isUpdateCartItemError, updateCartItemError]);

  const handleIncreaseQuantity = useCallback(() => {
    setQuantity((prev) => prev + 1);
  }, []);

  const handleDecreaseQuantity = useCallback(() => {
    setQuantity((prev) => {
      if (prev > 1) {
        return prev - 1;
      }
      return 1;
    });
  }, []);

  const handleSubmit = useCallback(() => {
    if (cartItem) {
      updateCartItem({ cartItemId: cartItem.id, quantity });
    } else {
      addToCart({ productId, quantity });
    }
  }, [productId, quantity, cartItem]);

  return (
    <View style={styles.footer}>
      <View style={styles.quantity}>
        <Typography text="Quantity" size={SizeType.Small} />
        <View style={styles.quantityCount}>
          <TouchableOpacity
            hitSlop={{ top: 10, left: 10, right: 10, bottom: 10 }}
            style={styles.quantityItem}
            onPress={handleDecreaseQuantity}
            testID="decrease-quantity"
          >
            <MinusIcon />
          </TouchableOpacity>
          <View style={[styles.quantityItem, styles.quantityNumber]}>
            <Typography
              fontWeight={FontWeight.Medium}
              text={quantity}
              size={SizeType.Large}
              variant={TypoVariant.Paragraph2}
            />
          </View>
          <TouchableOpacity
            hitSlop={{ top: 10, left: 10, right: 10, bottom: 10 }}
            style={styles.quantityItem}
            onPress={handleIncreaseQuantity}
            testID="add-quantity"
          >
            <AddIcon />
          </TouchableOpacity>
        </View>
      </View>
      <Button
        disabled={isAddCartPending || isUpdateCartPending}
        title={cartItem ? 'Update' : 'Add to cart'}
        icon={IconName.Cart}
        onPress={handleSubmit}
      />
    </View>
  );
};

export default memo(ProductFooter);
