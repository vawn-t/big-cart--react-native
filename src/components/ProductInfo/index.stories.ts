import type { Meta, StoryObj } from '@storybook/react-native';

import ProductDetail from '.';

const meta: Meta<typeof ProductDetail> = {
  title: 'Components/ProductDetail',
  component: ProductDetail,
};

export default meta;

type Story = StoryObj<React.ComponentProps<typeof ProductDetail>>;

export const Base: Story = {
  args: {
    id: 1,
    description:
      'Organic Mountain works as a seller for many organic growers of organic lemons. Organic lemons are easy to spot in your produce aisle. They are just like regular lemons, but they will usually have a few more scars on the outside of the lemon skin. Organic lemons are considered to be the world`s finest lemon for juicing',
    productUnitId: 1,
    price: 2.22,
    name: 'Organic Lemons',
  },
};
