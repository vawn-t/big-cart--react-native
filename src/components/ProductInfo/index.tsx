import { View } from 'react-native';

// Components
import ProductContent from './ProductContent';
import ProductFooter from './ProductFooter';

// Styles
import styles from './styles';

interface IProps {
  id: number;
  price: number;
  name: string;
  productUnitId: number;
  description: string;
}

const ProductInfo = ({
  id,
  price,
  name,
  productUnitId,
  description,
}: IProps) => (
  <View style={styles.content}>
    <ProductContent
      price={price}
      name={name}
      productUnitId={productUnitId}
      description={description}
    />
    <ProductFooter productId={id} />
  </View>
);

export default ProductInfo;
