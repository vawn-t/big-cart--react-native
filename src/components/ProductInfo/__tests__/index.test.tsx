import ProductInfo from '..';
import { render } from '@utils/index';

jest.mock('@react-native-async-storage/async-storage', () =>
  require('@react-native-async-storage/async-storage/jest/async-storage-mock'),
);

const ownProps: React.ComponentProps<typeof ProductInfo> = {
  id: 1,
  price: 10,
  name: 'Sample Product',
  productUnitId: 1,
  description: 'A sample product description',
};

const setup = (overrideProps = {}) => {
  const props = { ...ownProps, ...overrideProps };

  return render(<ProductInfo {...props} />);
};

describe('ProductInfo component', () => {
  it('renders correctly', () => {
    const { toJSON } = setup({
      id: 1,
      price: 10,
      name: 'Sample Product',
      amount: '2kg',
      description: 'A sample product description',
      numberInCart: 3,
    });
    expect(toJSON()).toMatchSnapshot();
  });
});
