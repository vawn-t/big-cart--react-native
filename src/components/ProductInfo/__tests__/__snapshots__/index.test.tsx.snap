// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductInfo component renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNSScreenStack
    style={
      {
        "flex": 1,
      }
    }
  >
    <RNSScreen
      collapsable={false}
      gestureResponseDistance={
        {
          "bottom": -1,
          "end": -1,
          "start": -1,
          "top": -1,
        }
      }
      nativeBackButtonDismissalEnabled={false}
      onAppear={[Function]}
      onDisappear={[Function]}
      onDismissed={[Function]}
      onHeaderBackButtonClicked={[Function]}
      onNativeDismissCancelled={[Function]}
      onTransitionProgress={[Function]}
      onWillDisappear={[Function]}
      replaceAnimation="push"
      sheetAllowedDetents="large"
      sheetCornerRadius={-1}
      sheetExpandsWhenScrolledToEdge={true}
      sheetGrabberVisible={false}
      sheetLargestUndimmedDetent="all"
      stackPresentation="push"
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      swipeDirection="horizontal"
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        style={
          {
            "flex": 1,
            "flexDirection": "column-reverse",
          }
        }
      >
        <View
          collapsable={false}
          style={
            [
              {
                "flex": 1,
              },
              {
                "backgroundColor": "rgb(242, 242, 242)",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#F4F5F9",
                "borderTopEndRadius": 10,
                "borderTopStartRadius": 10,
                "flex": 1,
                "paddingHorizontal": 15,
                "paddingVertical": 20,
                "rowGap": 25,
              }
            }
          >
            <View
              style={
                {
                  "alignSelf": "flex-start",
                  "rowGap": 5,
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "color": "#868889",
                    },
                    {
                      "fontFamily": "Poppins-SemiBold",
                    },
                    {
                      "fontSize": 15,
                    },
                    {
                      "color": "#6CC51D",
                    },
                  ]
                }
              >
                $10
              </Text>
              <Text
                style={
                  [
                    {
                      "color": "#000000",
                    },
                    {
                      "fontFamily": "Poppins-SemiBold",
                    },
                    {
                      "fontSize": 20,
                    },
                    undefined,
                  ]
                }
              >
                Sample Product
              </Text>
              <Text
                style={
                  [
                    {
                      "color": "#868889",
                    },
                    {
                      "fontFamily": "Poppins-Medium",
                    },
                    {
                      "fontSize": 12,
                    },
                    undefined,
                  ]
                }
              />
            </View>
            <Text
              numberOfLines={6}
              style={
                [
                  {
                    "color": "#868889",
                  },
                  {
                    "fontFamily": "Poppins-Regular",
                  },
                  {
                    "fontSize": 12,
                  },
                  {
                    "lineHeight": 18,
                    "width": "100%",
                  },
                ]
              }
            >
              A sample product description
            </Text>
            <View
              style={
                {
                  "bottom": 40,
                  "position": "absolute",
                  "rowGap": 10,
                  "width": "100%",
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#FFFFFF",
                    "borderRadius": 5,
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                    "paddingLeft": 15,
                  }
                }
              >
                <Text
                  style={
                    [
                      {
                        "color": "#868889",
                      },
                      {
                        "fontFamily": "Poppins-Regular",
                      },
                      {
                        "fontSize": 12,
                      },
                      undefined,
                    ]
                  }
                >
                  Quantity
                </Text>
                <View
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                    }
                  }
                >
                  <View
                    accessibilityState={
                      {
                        "busy": undefined,
                        "checked": undefined,
                        "disabled": undefined,
                        "expanded": undefined,
                        "selected": undefined,
                      }
                    }
                    accessibilityValue={
                      {
                        "max": undefined,
                        "min": undefined,
                        "now": undefined,
                        "text": undefined,
                      }
                    }
                    accessible={true}
                    collapsable={false}
                    focusable={true}
                    hitSlop={
                      {
                        "bottom": 10,
                        "left": 10,
                        "right": 10,
                        "top": 10,
                      }
                    }
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      {
                        "alignItems": "center",
                        "justifyContent": "center",
                        "opacity": 1,
                        "width": 48,
                      }
                    }
                    testID="decrease-quantity"
                  >
                    <RNSVGSvgView
                      bbHeight="24"
                      bbWidth="24"
                      fill="none"
                      focusable={false}
                      height={24}
                      style={
                        [
                          {
                            "backgroundColor": "transparent",
                            "borderWidth": 0,
                          },
                          {
                            "flex": 0,
                            "height": 24,
                            "width": 24,
                          },
                        ]
                      }
                      width={24}
                    >
                      <RNSVGGroup
                        fill={null}
                        propList={
                          [
                            "fill",
                          ]
                        }
                      >
                        <RNSVGPath
                          d="M6 12h12"
                          fill={
                            {
                              "payload": 4278190080,
                              "type": 0,
                            }
                          }
                          propList={
                            [
                              "stroke",
                              "strokeWidth",
                              "strokeLinecap",
                              "strokeLinejoin",
                            ]
                          }
                          stroke={
                            {
                              "payload": 4285318429,
                              "type": 0,
                            }
                          }
                          strokeLinecap={1}
                          strokeLinejoin={1}
                          strokeWidth="1.5"
                        />
                      </RNSVGGroup>
                    </RNSVGSvgView>
                  </View>
                  <View
                    style={
                      [
                        {
                          "alignItems": "center",
                          "justifyContent": "center",
                          "width": 48,
                        },
                        {
                          "borderColor": "#EBEBEB",
                          "borderLeftWidth": 1,
                          "borderRightWidth": 1,
                          "paddingVertical": 10,
                        },
                      ]
                    }
                  >
                    <Text
                      style={
                        [
                          {
                            "color": "#000000",
                          },
                          {
                            "fontFamily": "Poppins-Medium",
                          },
                          {
                            "fontSize": 20,
                          },
                          undefined,
                        ]
                      }
                    >
                      1
                    </Text>
                  </View>
                  <View
                    accessibilityState={
                      {
                        "busy": undefined,
                        "checked": undefined,
                        "disabled": undefined,
                        "expanded": undefined,
                        "selected": undefined,
                      }
                    }
                    accessibilityValue={
                      {
                        "max": undefined,
                        "min": undefined,
                        "now": undefined,
                        "text": undefined,
                      }
                    }
                    accessible={true}
                    collapsable={false}
                    focusable={true}
                    hitSlop={
                      {
                        "bottom": 10,
                        "left": 10,
                        "right": 10,
                        "top": 10,
                      }
                    }
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      {
                        "alignItems": "center",
                        "justifyContent": "center",
                        "opacity": 1,
                        "width": 48,
                      }
                    }
                    testID="add-quantity"
                  >
                    <RNSVGSvgView
                      bbHeight="24"
                      bbWidth="24"
                      fill="none"
                      focusable={false}
                      height={24}
                      style={
                        [
                          {
                            "backgroundColor": "transparent",
                            "borderWidth": 0,
                          },
                          {
                            "flex": 0,
                            "height": 24,
                            "width": 24,
                          },
                        ]
                      }
                      width={24}
                    >
                      <RNSVGGroup
                        fill={null}
                        propList={
                          [
                            "fill",
                          ]
                        }
                      >
                        <RNSVGPath
                          d="M6 12h12M12 18V6"
                          fill={
                            {
                              "payload": 4278190080,
                              "type": 0,
                            }
                          }
                          propList={
                            [
                              "stroke",
                              "strokeWidth",
                              "strokeLinecap",
                              "strokeLinejoin",
                            ]
                          }
                          stroke={
                            {
                              "payload": 4285318429,
                              "type": 0,
                            }
                          }
                          strokeLinecap={1}
                          strokeLinejoin={1}
                          strokeWidth="1.5"
                        />
                      </RNSVGGroup>
                    </RNSVGSvgView>
                  </View>
                </View>
              </View>
              <View
                accessibilityState={
                  {
                    "disabled": false,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  [
                    {
                      "alignItems": "center",
                      "backgroundColor": "#6CC51D",
                      "borderRadius": 5,
                      "flexDirection": "row",
                    },
                    null,
                    undefined,
                  ]
                }
              >
                <View
                  style={
                    {
                      "flexGrow": 1,
                      "paddingVertical": 20,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "alignItems": "flex-start",
                        "bottom": 0,
                        "justifyContent": "center",
                        "left": 30,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                      }
                    }
                  >
                    <RNSVGSvgView
                      bbHeight="25"
                      bbWidth="21"
                      fill="none"
                      focusable={false}
                      height={25}
                      style={
                        [
                          {
                            "backgroundColor": "transparent",
                            "borderWidth": 0,
                          },
                          {
                            "flex": 0,
                            "height": 25,
                            "width": 21,
                          },
                        ]
                      }
                      testID="Button-icon"
                      width={21}
                    >
                      <RNSVGGroup
                        fill={null}
                        propList={
                          [
                            "fill",
                          ]
                        }
                      >
                        <RNSVGPath
                          d="M20.995 21.61 19.492 5.49c-.032-.357-.348-.628-.719-.628h-3.09C15.64 2.172 13.333 0 10.5 0 7.667 0 5.36 2.173 5.317 4.862h-3.09a.707.707 0 0 0-.72.629L.006 21.61c0 .02-.005.04-.005.06C0 23.508 1.765 25 3.938 25h13.124C19.235 25 21 23.507 21 21.672c0-.02 0-.041-.005-.062ZM10.5 1.38c2.034 0 3.691 1.555 3.734 3.482H6.766c.043-1.927 1.7-3.482 3.734-3.482Zm6.562 22.24H3.938c-1.363 0-2.468-.86-2.49-1.918L2.888 6.248h2.425v2.097c0 .383.322.69.724.69.402 0 .724-.307.724-.69V6.247h7.474v2.097c0 .383.322.69.725.69.402 0 .724-.307.724-.69V6.247h2.425l1.443 15.456c-.021 1.058-1.132 1.917-2.49 1.917Z"
                          fill={
                            {
                              "payload": 4294967295,
                              "type": 0,
                            }
                          }
                          propList={
                            [
                              "fill",
                            ]
                          }
                        />
                      </RNSVGGroup>
                    </RNSVGSvgView>
                  </View>
                  <Text
                    style={
                      [
                        {
                          "color": "#000000",
                        },
                        {
                          "fontFamily": "Poppins-SemiBold",
                        },
                        {
                          "fontSize": 15,
                        },
                        {
                          "color": "#FFFFFF",
                          "textAlign": "center",
                        },
                      ]
                    }
                  >
                    Add to cart
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
      <RNSScreenStackHeaderConfig
        backButtonInCustomView={false}
        backTitleVisible={true}
        backgroundColor="rgb(255, 255, 255)"
        color="rgb(0, 122, 255)"
        direction="ltr"
        disableBackButtonMenu={false}
        hidden={false}
        hideBackButton={false}
        largeTitleHideShadow={false}
        title="test"
        titleColor="rgb(28, 28, 30)"
        topInsetEnabled={false}
        translucent={false}
      />
    </RNSScreen>
  </RNSScreenStack>
</RNCSafeAreaProvider>
`;
